autopush_rs/src/util/mod.rs

```diff
@@ -41,52 +44,50 @@ where
     }))
 }
 
-static LOG_JSON: AtomicBool = ATOMIC_BOOL_INIT;
 
-pub fn init_logging(json: bool) {
-    // We only initialize once, so ignore initialization errors related to
-    // calling this function twice.
-    let mut builder = env_logger::LogBuilder::new();
-
-    if env::var("RUST_LOG").is_ok() {
-        builder.parse(&env::var("RUST_LOG").unwrap());
-    }
-
-    builder.target(env_logger::LogTarget::Stdout).format(
-        maybe_json_record,
-    );
-
-    if builder.init().is_ok() {
-        LOG_JSON.store(json, Ordering::SeqCst);
-    }
+// Hold a reference to the log guards for scoped logging which requires these to stay alive
+// for the implicit logger to be passed into logging calls
+pub struct LogGuards {
+    _scope_guard: slog_scope::GlobalLoggerGuard,
+    _log_guard: (),
 }
 
-fn maybe_json_record(record: &LogRecord) -> String {
-    #[derive(Serialize)]
-    struct Message<'a> {
-        msg: String,
-        level: String,
-        target: &'a str,
-        module: &'a str,
-        file: &'a str,
-        line: u32,
-    }
-
-    if LOG_JSON.load(Ordering::SeqCst) {
-        serde_json::to_string(&Message {
-            msg: record.args().to_string(),
-            level: record.level().to_string(),
-            target: record.target(),
-            module: record.location().module_path(),
-            file: record.location().file(),
-            line: record.location().line(),
-        }).unwrap()
+pub fn init_logging(json: bool, hostname: &'static str) -> LogGuards {
+    let decorator = slog_term::TermDecorator::new().build();
+    if json {
+        let drain = autojson::AutoJson::new(io::stdout())
+            .add_default_keys()
+            .add_key_value(o!(
+                "Hostname" => hostname,
+                "Type" => "autopush_rs:log",
+                "EnvVersion" => "2.0",
+                "Logger" => format!("AutopushRust-{}", env!("CARGO_PKG_VERSION")),
+            ))
+            .build()
+            .fuse();
+        let drain = slog_async::Async::new(drain).build().fuse();
+        let logger = slog::Logger::root(drain, o!());
+        let _scope_guard = slog_scope::set_global_logger(logger);
+        let _log_guard = match slog_stdlog::init() {
```

Comment:
what's with _log_guard? it's always () -- init returns an empty Result https://docs.rs/slog-stdlog/1.1.0/slog_stdlog/fn.init.html

Reply:
You might want to propagate the Error from slog init (have init_logging return a Result) then the init_logging caller can expect() on it, like it's doing with all the other config values

Reply:
The stdlog init only errors out if its called twice due to a global it has. This only happens with our tests, so I ignored the error so our tests pass.

Reply:
Ok, ate the result.